postgresql/database_test.go

```diff
@@ -1912,3 +1913,50 @@ func TestOptionTypeJsonbStruct(t *testing.T) {
 		t.Fatalf("Expecting Num to be 123")
 	}
 }
+
+// TestExhaustConnections simulates a "too many connections" situation
+// triggered by opening more transactions than available connections.
+// upper.io/db deals with this problem by waiting a bit more for the connection
+// to be established.
+func TestExhaustConnections(t *testing.T) {
+	var err error
+	var sess db.Database
+	var wg sync.WaitGroup
+
+	if sess, err = db.Open(Adapter, settings); err != nil {
+		t.Fatal(err)
+	}
+
+	// By default, PostgreSQL accepts 100 connections only.
+	for i := 0; i < 500; i++ {
+		wg.Add(1)
+		t.Logf("Tx %d: Pending", i)
+		go func(t *testing.T, wg *sync.WaitGroup, i int) {
+			var tx db.Tx
+			defer wg.Done()
+
+			start := time.Now()
+
+			// Requesting a new transaction session.
+			if tx, err = sess.Transaction(); err != nil {
```

Comment:
This `Transaction()` will wait on its goroutine until a connection is available, this waiting time ranges from immediately up to 5s (it depends on `waitForConnection()`)